fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/cursors/AutoContinuingCursorTest.java

```diff
@@ -54,28 +64,82 @@ public void getDatabase() {
+            int retryAttempts,
+            List<Integer> expectedList) {
         try (FDBDatabaseRunner runner = database.newRunner()) {
-            RecordCursor<Integer> cursor = new AutoContinuingCursor<>(runner, nextCursorGenerator);
+            RecordCursor<Integer> cursor = new AutoContinuingCursor<>(runner, nextCursorGenerator, retryAttempts);
 
             List<Integer> returnedList = cursor.asList().join();
             assertEquals(expectedList, returnedList);
         }
     }
 
     @Test
-    public void testAutoContinuingCursorSimple() {
+    void testAutoContinuingCursorSimple() {
         testAutoContinuingCursorGivenCursorGenerator((context, continuation) ->
                         new ListCursor<>(list, continuation).limitRowsTo(3),
                 list);
     }
 
     @Test
-    public void testAutoContinuingCursorWhenSomeGeneratedCursorsNeverHaveNext() {
+    void testAutoContinuingCursorWhenSomeGeneratedCursorsNeverHaveNext() {
         // This underlying cursor may not produce any item in one transaction. AutoContinuingCursor is expected to make
         // progress until it is truly exhausted.
         testAutoContinuingCursorGivenCursorGenerator((context, continuation) ->
                         new ListCursor<>(list, continuation).limitRowsTo(2).filter(item -> item % 3 == 0),
                 Arrays.asList(3, 6, 9)
         );
     }
+
+    @Test
+    void testContinuesOnRetryableException() {
+        final AtomicInteger iteration = new AtomicInteger(0);
+
+        testAutoContinuingCursorGivenCursorGenerator((context, continuation) ->
+                new TestingListCursor<>(list, continuation, () -> {
+                    if (iteration.incrementAndGet() % 3 == 0) {
+                        CompletableFuture<Void> failedFuture = new CompletableFuture<>();
+                        failedFuture.completeExceptionally(new FDBException("transaction_too_old", FDBError.TRANSACTION_TOO_OLD.code()));
+                        return failedFuture;
+                    }
+                    return AsyncUtil.DONE;
+                }), 1, list);
+    }
+
+    @Test
+    void testRetryableMaxAttempts() {
+        final AtomicInteger exceptionCount = new AtomicInteger(0);
+
+        CompletionException e = assertThrows(CompletionException.class, () ->
+            testAutoContinuingCursorGivenCursorGenerator((context, continuation) ->
+                    new TestingListCursor<>(list, continuation, () -> {
+                        exceptionCount.incrementAndGet();
+                        CompletableFuture<Void> failedFuture = new CompletableFuture<>();
+                        failedFuture.completeExceptionally(new FDBException("transaction_too_old", FDBError.TRANSACTION_TOO_OLD.code()));
+                        return failedFuture;
+                    }), 3, list));
+        assertEquals(((FDBException) e.getCause()).getCode(), FDBError.TRANSACTION_TOO_OLD.code());
+        assertEquals(exceptionCount.get(), 4);
```

Comment:
If the max attempts is 3 but the exception count is 4, does that indicate there's an off-by-one error somewhere? (Or that the parameter needs to be renamed to "max retries" instead of "max attempts"?)